rancher/rancher
# Pin Fleet version

## Issue: [fleet#1590](https://github.com/rancher/fleet/issues/1590)
 
## Problem

There are situations where Rancher would update Fleet to its latest available version, without that upgrade being triggered by a Rancher update. This causes confusion among users, with the `CATTLE_FLEET_MIN_VERSION` environment variable not always reflecting the actual installed Fleet version.
 
## Solution
This adds a new `CATTLE_FLEET_VERSION` environment variable, similarly to what has been done for the `rancher-webhook` chart starting with #41302.
The corresponding exact version setting for Fleet is monitored, so that a change leads to Fleet being updated _to that exact version_.
Contrarily to behaviour implemented for `rancher-webhook` though, Fleet may still be updated manually via the Rancher UI without that update being undone via a periodical check: for Fleet, Rancher restores the chart to its pinned version only if that pinned version is more recent than the currently installed version.

### Backward compatibility: precedence of min version
Pre-existing setting `CATTLE_FLEET_MIN_VERSION` is now deprecated in favour of `CATTLE_FLEET_VERSION`, and kept for backward compatibility. If `CATTLE_FLEET_MIN_VERSION` is set, it has precedence over `CATTLE_FLEET_VERSION` as the version of Fleet to be installed.

### New min version behaviour
Whichever setting the version of Fleet to be installed comes from, that version is considered a _min version_, in the sense that it will not overwrite an already installed, _newer_, version of Fleet. However, in no case should the _latest available_ Fleet version be installed by default. This would only happen if none of `CATTLE_FLEET_MIN_VERSION` or `CATTLE_FLEET_VERSION` is specified, which would be symptomatic of a build/configuration error.

Fleet being the only known Rancher component using min version behaviour, this change is not expected to impact any other component.
 
## Testing
## Engineering Testing
### Manual Testing
Started Rancher on my machine via [these steps](https://github.com/rancher/rancher/wiki/Setting-Up-Rancher-2.0-Development-Environment#setting-up-for-rancher-development) and monitored Fleet install behaviour, with the following 2 tweaks:
* pointed Rancher, via environment variables, to a [test branch](https://github.com/rancher/charts/tree/test-tan) of `rancher/charts` on which I could remove and restore version `0.9.0-rc.3` of Fleet charts.
* hard-coded `SystemFeatureChartRefreshSeconds` as 30s (instead of 6 hours) locally to speed up testing

#### Before (using branch `release/v2.8`)
1. Started Rancher with `CATTLE_FLEET_MIN_VERSION` set to `102.1.1+up0.7.1`, with `0.9.0-rc.3` being in the relevant `rancher/charts` branch
2. Could see Fleet charts being installed in version `0.9.1-rc.3` 

#### After (using branch `fleet-pin-version`)
1. Started Rancher with `CATTLE_FLEET_VERSION` set to `102.1.1+up0.7.1` and  `0.9.0-rc.3` being in the relevant `rancher/charts` branch
2. Could see Fleet charts being installed in version `0.7.1` 

### Automated Testing
* Test types added/modified:
    * Unit tests for:
        * Fleet charts controller: checking for changes in Fleet version settings triggering installation of Fleet charts
        * System manager called by said controller: checking that chart installation happens or not for combinations of pinned, already deployed and requested exact versions of Fleet
    * Integration (Go Framework): Added settings test case for exact Fleet version (disclaimer: did not run it)

## QA Testing Considerations
* The following use cases must be validated:

| Exact version (`CATTLE_FLEET_VERSION`) | Min version (`CATTLE_FLEET_MIN_VERSION`) | Result |
| ------------- | ------------- | ------------- |
| Version present in charts catalog | Empty | `CATTLE_FLEET_VERSION` Fleet version installed |
| Version X present in charts catalog | Version Y present in charts catalog | `CATTLE_FLEET_MIN_VERSION` Fleet version installed |
| Empty | Version not present in charts catalog | Error (should never happen, build/config error) |
| Empty | Version present in charts catalog | `CATTLE_FLEET_MIN_VERSION` Fleet version installed |
| Update via the Rancher web UI (see [comment](https://github.com/rancher/rancher/pull/42304#discussion_r1297281360)) with newer version selected than already installed | Version present in catalog | Selected version is installed |
| Update via the Rancher web UI (see [comment](https://github.com/rancher/rancher/pull/42304#discussion_r1297281360)) with older version selected than already installed | Version present in catalog | Selected version is not installed; already installed version is kept |

* Updating to 2.8.x from an old 2.8.(x-1) should lead to the `CATTLE_FLEET_VERSION` for Rancher 2.8.x being installed.
    * Note: this use case cannot be tested as long as a _single_ Rancher release exists with this Fleet version pinning functionality
